test/e2e/resource_quota.go

```diff
@@ -41,6 +41,79 @@ const (
+		By("Creating a ResourceQuota")
+		quotaName := "test-quota"
+		resourceQuota := newTestResourceQuotaForEphemeralStorage(quotaName)
+		resourceQuota, err := createResourceQuota(f.ClientSet, f.Namespace.Name, resourceQuota)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Ensuring resource quota status is calculated")
+		usedResources := v1.ResourceList{}
+		usedResources[v1.ResourceQuotas] = resource.MustParse("1")
+		err = waitForResourceQuota(f.ClientSet, f.Namespace.Name, quotaName, usedResources)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Creating a Pod that fits quota")
+		podName := "test-pod"
+		requests := v1.ResourceList{}
+		requests[v1.ResourceEphemeralStorage] = resource.MustParse("300Mi")
+		pod := newTestPodForQuota(f, podName, requests, v1.ResourceList{})
+		pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(pod)
+		Expect(err).NotTo(HaveOccurred())
+		podToUpdate := pod
+
+		By("Ensuring ResourceQuota status captures the pod usage")
+		usedResources[v1.ResourceQuotas] = resource.MustParse("1")
+		usedResources[v1.ResourcePods] = resource.MustParse("1")
+		usedResources[v1.ResourceEphemeralStorage] = requests[v1.ResourceEphemeralStorage]
+		err = waitForResourceQuota(f.ClientSet, f.Namespace.Name, quotaName, usedResources)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Not allowing a pod to be created that exceeds remaining quota")
+		requests = v1.ResourceList{}
+		requests[v1.ResourceEphemeralStorage] = resource.MustParse("300Mi")
+		pod = newTestPodForQuota(f, "fail-pod", requests, v1.ResourceList{})
+		pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(pod)
+		Expect(err).To(HaveOccurred())
+
+		By("Ensuring a pod cannot update its resource requirements")
+		// a pod cannot dynamically update its resource requirements.
+		requests = v1.ResourceList{}
+		requests[v1.ResourceEphemeralStorage] = resource.MustParse("100Mi")
+		podToUpdate.Spec.Containers[0].Resources.Requests = requests
+		_, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Update(podToUpdate)
+		Expect(err).To(HaveOccurred())
+
+		By("Ensuring attempts to update pod resource requirements did not change quota usage")
+		err = waitForResourceQuota(f.ClientSet, f.Namespace.Name, quotaName, usedResources)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Deleting the pod")
+		err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete(podName, metav1.NewDeleteOptions(0))
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Ensuring resource quota status released the pod usage")
+		usedResources[v1.ResourceQuotas] = resource.MustParse("1")
+		usedResources[v1.ResourcePods] = resource.MustParse("0")
+		usedResources[v1.ResourceEphemeralStorage] = resource.MustParse("0")
+		err = waitForResourceQuota(f.ClientSet, f.Namespace.Name, quotaName, usedResources)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Removing resourceQuota")
+		deleteResourceQuota(f.ClientSet, f.Namespace.Name, resourceQuota.Name)
+		Expect(err).NotTo(HaveOccurred())
+	})
+})
```

Comment:
About deleting resources created by the test suite: since the test could fail at any point after a resource is created, and we want the test to cleanup its environment explicitly (rather than relying on the generated namespace being deleted), a `defer` after the resource is created or an `AfterEach` in the `Context` (or `Describe`) are commonly used for this purpose.

Reply:
I found we did the deletion in the test in other places like resourcequota(cpu,mem part).
If we want to delete all the pods and other resources in AfterEach function, should we use different var names to represent the pods created in the test So that we can delete all of them?

Reply:
If I understand your question, you need the pod's name to Delete it. This can be done in an `AfterEach` but then you'll need the scope of the `podName` variable to be global to the parent `Context` or `Describe`. If you delete the pod in a `defer` then you can create the pod value in the `It` block and defer-delete it after the pod-Create is successful.
If I understand your question, you can delete a pod in an `AfterEach` but you'll need the scope of the `podName` variable to be global to the parent `Context` or `Describe`. If you delete the pod in a `defer` then you create the pod value in the `It` block and defer-delete it after the pod-Create is successful.

Reply:
done